test/gtest/uct/test_wakeup.cc

```diff
@@ -101,3 +108,257 @@ UCS_TEST_P(test_uct_wakeup, am)
+    ASSERT_EQ(status, UCS_OK);
+
+    /* make sure [send|recv]_cq handled properly */
+    check_send_cq(m_e1->iface(), 1);
+    check_recv_cq(m_e1->iface(), 0);
+}
+
+
+UCS_TEST_P(test_uct_wakeup_ib, txrx_cq)
+{
+    ucs_status_t status;
+
+    initialize();
+
+    status = uct_wakeup_efd_arm(wakeup_handle);
+    ASSERT_EQ(status, UCS_OK);
+
+    /* check initial state of the fd and [send|recv]_cq */
+    EXPECT_EQ(poll(&wakeup_fd, 1, 0), 0);
+    check_send_cq(m_e1->iface(), 0);
+    check_recv_cq(m_e1->iface(), 0);
+
+    /* send the data */
+    send_msg_e1_e2();
+    send_msg_e2_e1();
+
+    twait(50); /* Let completion to be generated */
+
+    /* make sure the file descriptor is signaled */
+    ASSERT_EQ(poll(&wakeup_fd, 1, 1), 1);
+
+    status = uct_wakeup_wait(wakeup_handle);
+    ASSERT_EQ(status, UCS_OK);
+
+    /* make sure [send|recv]_cq handled properly */
+    check_send_cq(m_e1->iface(), 1);
+    check_recv_cq(m_e1->iface(), 1);
+
+}
+
+
+UCS_TEST_P(test_uct_wakeup_ib, multi_txrx_cq)
+{
+    const size_t msg_count = 5;
+    ucs_status_t status;
+
+    initialize();
+
+    status = uct_wakeup_efd_arm(wakeup_handle);
+    ASSERT_EQ(status, UCS_OK);
+
+    /* check initial state of the fd and [send|recv]_cq */
+    EXPECT_EQ(poll(&wakeup_fd, 1, 0), 0);
+    check_send_cq(m_e1->iface(), 0);
+    check_recv_cq(m_e1->iface(), 0);
+
+    /* send the data */
+    send_msg_e1_e2(msg_count);
+    send_msg_e2_e1(msg_count);
+
+    /* Make sure all messages delivered */
+    while ((test_uct_wakeup::am_handler_count    < msg_count) ||
+           (test_uct_wakeup_ib::bcopy_pack_count < msg_count)) {
+        progress();
+    }
+
+    /* make sure the file descriptor is signaled */
+    ASSERT_EQ(poll(&wakeup_fd, 1, 1), 1);
+
+    status = uct_wakeup_wait(wakeup_handle);
+    ASSERT_EQ(status, UCS_OK);
+
+    /* make sure the file descriptor is cleared */
+    ASSERT_EQ(poll(&wakeup_fd, 1, 1), 0);
+
+    /* make sure [send|recv]_cq handled properly */
+    check_send_cq(m_e1->iface(), 1);
+    check_recv_cq(m_e1->iface(), 1);
+
+}
+
+
+UCS_TEST_P(test_uct_wakeup_ib, wait_after_arm)
+{
+    const size_t msg_count = 2;
+    ucs_status_t status;
+
+    initialize();
+
+    status = uct_wakeup_efd_arm(wakeup_handle);
+    ASSERT_EQ(status, UCS_OK);
+
+    /* check initial state of the fd */
+    EXPECT_EQ(poll(&wakeup_fd, 1, 0), 0);
+
+    /* send the data */
+    send_msg_e1_e2(msg_count);
+    send_msg_e2_e1(msg_count);
+
+    /* Make sure all messages delivered */
+    while ((test_uct_wakeup::am_handler_count    < msg_count) ||
+           (test_uct_wakeup_ib::bcopy_pack_count < msg_count)) {
+        progress();
+    }
+
+    status = uct_wakeup_efd_arm(wakeup_handle);
+    ASSERT_EQ(status, UCS_OK);
+
+    /** make sure the file descriptor is not signaled
+     * If the msg_count is 1 or avoid last wakeup_arm()
+     * the fd will be signaled anyway
+     */
+    ASSERT_EQ(poll(&wakeup_fd, 1, 1), 0);
```

Comment:
@yosefe This `fd` is not signalled because `uct_wakeup_efd_arm()` called before. If we use `msg_count == 1`, the `fd` is signalled. If we remove `uct_wakeup_efd_arm()` before, the `fd` will be signalled also.
Do you have an idea why?